crypto/slh_dsa/slh_dsa.c

```diff
@@ -0,0 +1,359 @@
+        return 0;
+    }
+
+    /* Exit if signature is invalid size */
+    if (sig_len != params->sig_len
+            || !PACKET_buf_init(sig_rpkt, sig, sig_len))
+        return 0;
+    if (!PACKET_get_bytes(sig_rpkt, &r, n))
+        return 0;
+
+    adrsf->zero(adrs);
+
+    pk_seed = SLH_DSA_PK_SEED(pub);
+    pk_root = SLH_DSA_PK_ROOT(pub);
+
+    if (!hashf->H_MSG(hctx, r, pk_seed, pk_root, msg, msg_len,
+                      m_digest, sizeof(m_digest)))
+        return 0;
+
+    /*
+     * Get md (the first md_len bytes of m_digest to use in
+     * ossl_slh_fors_pk_from_sig(), and then retrieve the tree id and leaf id
+     * from the remaining bytes in m_digest.
+     */
+    if (!PACKET_buf_init(m_digest_rpkt, m_digest, sizeof(m_digest))
+            || !PACKET_get_bytes(m_digest_rpkt, &md, md_len)
+            || !get_tree_ids(m_digest_rpkt, params, &tree_id, &leaf_id))
+        return 0;
+
+    adrsf->set_tree_address(adrs, tree_id);
+    adrsf->set_type_and_clear(adrs, SLH_ADRS_TYPE_FORS_TREE);
+    adrsf->set_keypair_address(adrs, leaf_id);
+    return ossl_slh_fors_pk_from_sig(hctx, sig_rpkt, md, pk_seed, adrs,
+                                     pk_fors, sizeof(pk_fors))
+        && ossl_slh_ht_verify(hctx, pk_fors, sig_rpkt, pk_seed,
+                              tree_id, leaf_id, pk_root)
+        && PACKET_remaining(sig_rpkt) == 0;
+}
+
+/**
+ * @brief Encode a message
+ * See FIPS 205 Algorithm 22 Step 8 (and algorithm 24 Step 4).
+ *
+ * SLH_DSA pure signatures are encoded as M' = 00 || ctx_len || ctx || msg
+ * Where ctx is the empty string by default and ctx_len <= 255.
+ *
+ * @param msg A message to encode
+ * @param msg_len The size of |msg|
+ * @param ctx An optional context to add to the message encoding.
+ * @param ctx_len The size of |ctx|. It must be in the range 0..255
+ * @param encode Use the Pure signature encoding if this is 1, and dont encode
+ *               if this value is 0.
+ * @param tmp A small buffer that may be used if the message is small.
+ * @param tmp_len The size of |tmp|
+ * @param out_len The size of the returned encoded buffer.
+ * @returns A buffer containing the encoded message. If the passed in
+ * |tmp| buffer is big enough to hold the encoded message then it returns |tmp|
+ * otherwise it allocates memory which must be freed by the caller. If |encode|
+ * is 0 then it returns |msg|. NULL is returned if there is a failure.
+ */
+static uint8_t *msg_encode(const uint8_t *msg, size_t msg_len,
+                           const uint8_t *ctx, size_t ctx_len, int encode,
+                           uint8_t *tmp, size_t tmp_len, size_t *out_len)
+{
+    uint8_t *encoded = NULL;
+    size_t encoded_len;
+
+    if (encode == 0) {
+        /* Raw message */
+        *out_len = msg_len;
+        return (uint8_t *)msg;
+    }
+    if (ctx_len > SLH_DSA_MAX_CONTEXT_STRING_LEN)
+        return NULL;
+
+    /* Pure encoding */
+    encoded_len = 1 + 1 + ctx_len + msg_len;
```

Comment:
it would be nice to note what the 1 + 1 is for  - sure I can see it below - but nice to not have to do that - so it is right in front of the reader.

Reply:
You can also see it above in the description of the function